dali/pipeline/data/tensor_vector.h

```diff
@@ -123,368 +92,110 @@ class TensorVector {
   }
 
   auto size() const noexcept {
-    return tensors_.size();
+    return curr_tensors_size_;
   }
 
   size_t ntensor() const noexcept {
-    return tensors_.size();
+    return curr_tensors_size_;
   }
 
-  size_t nbytes() const noexcept {
-    if (state_ == State::contiguous) {
-      return tl_->nbytes();
-    }
-    // else
-    size_t total_nbytes = 0;
-    for (const auto &t : tensors_) {
-      total_nbytes += t->nbytes();
-    }
-    return total_nbytes;
-  }
+  size_t nbytes() const noexcept;
+
+  size_t capacity() const noexcept;
+
+  TensorListShape<> shape() const;
 
-  size_t capacity() const noexcept {
-    if (state_ == State::contiguous) {
-      return tl_->capacity();
-    }
-    // else
-    size_t total_capacity = 0;
-    for (const auto &t : tensors_) {
-      total_capacity += t->capacity();
-    }
-    return total_capacity;
+  TensorShape<> tensor_shape(int idx) const {
+    assert(state_ == State::contiguous && curr_tensors_size_ == views_count_ ||
+           state_ == State::noncontiguous);
+    return tensors_[idx]->shape();
   }
 
-  TensorListShape<> shape() const {
-    if (state_ == State::contiguous) {
-      return tl_->shape();
-    }
-    if (tensors_.empty()) {
-      return {};
-    }
-    TensorListShape<> result(tensors_.size(), tensors_[0]->ndim());
-    for (size_t i = 0; i < tensors_.size(); i++) {
-      result.set_tensor_shape(i, tensors_[i]->shape());
-    }
-    return result;
+  const void *raw_tensor(int idx) const {
```

Comment:
Perhaps add `raw_mutable_tensor` for a complete API?
I know it would be a bit more involved, as it would potentially require updating the views (and most certainly we shouldn't do that on every call to `raw_mutable_tensor`).

Reply:
IMHO there should be an `assert` for this, but I don't have good idea how to do this...

Reply:
Ufff... I've checked that _raw_ _mutable_data does not silently allocate, so we're good whatever the state.